File: src/Python/134.GasStation.py

class Solution:
    def canCompleteCircuit(self, gas, cost):
        sum = 0
        total = 0
        start = 0
        for i in range(len(gas)):
            total += (gas[i] - cost[i])
            if sum < 0:
                sum = gas[i] - cost[i]
                start = i
            else:
                sum += (gas[i] - cost[i])
        return start if total >= 0 else -1 


if __name__ == "__main__":
    solution = Solution()
    print(solution.canCompleteCircuit([1,2,3,4,5], [3,4,5,1,2]))
    print(solution.canCompleteCircuit([2,3,4], [3,4,3]))
        

File: src/Java/101-200/134.GasStation.java

import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Queue;
import java.util.HashSet;
import java.util.LinkedList;

class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int total = 0;
        int nowsum = 0;
        int start = 0;
        for (int i = 0; i < gas.length; i++) {
            total += (gas[i] - cost[i]);
            if (nowsum < 0){
                nowsum = gas[i] - cost[i];
                start = i;
            }
            else
                nowsum += (gas[i] - cost[i]);
        }
        return total >= 0 ? start : -1;
    }
}

